gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java

```diff
@@ -127,16 +169,34 @@ public void putPathParam(Map<String, String> fields, String fieldName, Object fi
    * @param fieldValue a field value
    */
   public void putQueryParam(Map<String, List<String>> fields, String fieldName, Object fieldValue) {
-    ImmutableList.Builder<String> paramValueList = ImmutableList.builder();
+    ArrayList<String> paramValueList = new ArrayList();
     if (fieldValue instanceof List<?>) {
+      boolean hasProcessedMessage = false;
       for (Object fieldValueItem : (List<?>) fieldValue) {
-        paramValueList.add(String.valueOf(fieldValueItem));
+        if (isNonSerializableMessageValue(fieldValueItem)) {
+          putDecomposedMessageQueryParam(fields, fieldName, fieldValueItem);
+          hasProcessedMessage = true;
+        } else {
+          paramValueList.add(toQueryParamValue(fieldValueItem));
+        }
+      }
+      if (hasProcessedMessage) {
+        return;
       }
     } else {
-      paramValueList.add(String.valueOf(fieldValue));
+      if (isNonSerializableMessageValue(fieldValue)) {
+        putDecomposedMessageQueryParam(fields, fieldName, fieldValue);
+        return;
+      } else {
```

Comment:
When working with the pattern:
```
if (x is Collection)
  foreach i in x
    // do A(i)
else
  // do A(x)
```
We can often refactor to:
```java
Iterable<?> fieldValues =
    fieldValue instanceof Iterable
        ? (Iterable<?>) fieldValue
        : Collections.singleton(fieldValue);

foreach i in x
  // do A(i)
```
This eliminates the need to study both implementations of A to determine if they're the same or not, with the added benefit of it being easier to test the single code path rather than both. As the complexity of A increases, this simplification becomes more and more beneficial.